src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/FrozenDictionary.cs

```diff
@@ -105,14 +110,94 @@ public static class FrozenDictionary
-                    // the underlying FrozenHashtable's Int32 index as the keys themselves, avoiding the need to store the
-                    // same keys yet again.
-                    return typeof(TKey) == typeof(int) ?
-                        (FrozenDictionary<TKey, TValue>)(object)new Int32FrozenDictionary<TValue>((Dictionary<int, TValue>)(object)source) :
-                        new ValueTypeDefaultComparerFrozenDictionary<TKey, TValue>(source);
+#if NET7_0_OR_GREATER
+                    static FrozenDictionary<TKey, TValue> PickIntegerDictionary<TInt>(Dictionary<TKey, TValue> source)
+                        where TInt : struct, IBinaryInteger<TInt>
+                    {
+                        var k = (TInt[])(object)source.Keys.ToArray();
+                        var v = source.Values.ToArray();
+
+                        Array.Sort(k, v);
+
+                        var min = k[0];
+                        var max = k[k.Length - 1];
+                        var range = ulong.CreateTruncating(max - min);
+
+                        if (range == (ulong)k.Length - 1)
+                        {
+                            return (FrozenDictionary<TKey, TValue>)(object)new ClosedRangeIntegerFrozenDictionary<TInt, TValue>(k, v);
+                        }
+                        else if (k.Length <= MaxItemsInSmallIntegerDictionary)
+                        {
+                            return (FrozenDictionary<TKey, TValue>)(object)new SmallIntegerFrozenDictionary<TInt, TValue>(k, v);
+                        }
+                        else if (typeof(TInt) == typeof(int))
+                        {
+                            return (FrozenDictionary<TKey, TValue>)(object)new Int32FrozenDictionary<TValue>((Dictionary<int, TValue>)(object)source);
+                        }
+
+                        return new ValueTypeDefaultComparerFrozenDictionary<TKey, TValue>(source);
+                    }
+
+                    if (typeof(TKey) == typeof(int))
+                    {
+                        return PickIntegerDictionary<int>(source);
+                    }
+                    else if (typeof(TKey) == typeof(uint))
+                    {
+                        return PickIntegerDictionary<uint>(source);
+                    }
+                    else if (typeof(TKey) == typeof(long))
+                    {
+                        return PickIntegerDictionary<long>(source);
+                    }
+                    else if (typeof(TKey) == typeof(ulong))
+                    {
+                        return PickIntegerDictionary<ulong>(source);
+                    }
+                    else if (typeof(TKey) == typeof(short))
+                    {
+                        return PickIntegerDictionary<short>(source);
+                    }
+                    else if (typeof(TKey) == typeof(ushort))
+                    {
+                        return PickIntegerDictionary<ushort>(source);
+                    }
```

Comment:
Not byte and sbyte?

Reply:
Hmmm. Not sure how come I missed those. Done.